[bugfix] fix mamba slot leak when scheduling fails with radix cache (#15840)

kuafou opened the pull request:
## Motivation

This PR fixes a server crash caused by a false positive "memory leak detected" error when serving Mamba models with `radix-cache` enabled.

The issue was discovered during the investigation of #15840 but is a distinct issue regarding the scheduler's failure path cleanup. It occurs when the scheduler is temporarily resource-blocked (e.g., during request retraction), causing `check_memory()` to flag a temporarily untracked Mamba slot as leaked.

## Modifications

* Explicitly release the COW-allocated Mamba slot if `add_one_req` fails (returns `NO_TOKEN` or `IDLE`) during the scheduling phase.
* Add a regression test `test_mamba_slot_release_after_match_prefix_cow` to verify the cleanup logic.

## Accuracy Tests

N/A — resource cleanup only, no model output changes.

## Benchmarking and Profiling

N/A — failure-path fix, no performance impact.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### Details
**The Problem:**
1. In `init_next_round_input()`, a Mamba slot is allocated via Copy-On-Write (COW) / `match_prefix`.
2. Subsequently, `add_one_req()` is called but returns `NO_TOKEN` (e.g., because the KV cache pool is full).
3. The scheduler enters an "Idle" state (which technically means "resource-blocked" in this context) and triggers `self_check_during_idle()` -> `check_memory()`.
4. `check_memory()` fails because the slot allocated in step 1 is:
    * Not in `free_slots` (it was allocated).
    * Not yet tracked by `tree_cache` (the request didn't finish scheduling).
    * **Result:** The server crashes with `ValueError: token_to_kv_pool_allocator memory leak detected!`.

**The Fix:**
When scheduling fails, we now explicitly free the temporary Mamba slot. The request remains in the `waiting_queue` and will safely re-allocate (and copy) the Mamba state in the next scheduling round.

### Error Log
The server crashes with the following error during high load/retraction:

```text
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}

kuafou committed:
Fix mamba slot leak when scheduling fails with radix cache (#15840)

When add_one_req fails after init_next_round_input allocates a mamba slot
via COW (copy-on-write) during match_prefix, the slot was not released,
causing memory leak.

This fix releases the mamba slot when scheduling fails.

yizhang2077 commented on python/sglang/srt/managers/scheduler.py:
```diff
@@ -1982,6 +1982,12 @@ def get_new_batch_prefill(self) -> Optional[ScheduleBatch]:
 
             if res != AddReqResult.CONTINUE:
+                # Release mamba slot allocated via COW if scheduling fails
+                if self.is_hybrid_ssm and req.mamba_pool_idx is not None:
```
I think we need add condition for use mamba radix cache

kuafou committed:
Check tree_cache is MambaRadixCache before releasing COW mamba slot

COW (copy-on-write) mamba slot allocation only happens in MambaRadixCache.
Add isinstance check to ensure slot cleanup only runs when using MambaRadixCache,
following the pattern in scheduler_runtime_checker_mixin.py.

kuafou committed:
change comment

kuafou committed:
fix format

yizhang2077 commented:
/tag-and-rerun-ci

kuafou commented on python/sglang/srt/managers/scheduler.py:
```diff
@@ -1990,6 +1990,19 @@ def get_new_batch_prefill(self) -> Optional[ScheduleBatch]:
+                # check_memory() to detect a "memory leak" and crash the server.
+                # The next schedule round will re-allocate safely via match_prefix().
+                #
+                # See: https://github.com/sgl-project/sglang/issues/15840
```
Got it.

kuafou replied:
Done — the issue link has been removed and the PR is updated to the latest main.

kuafou committed:
remove issue link

kuafou committed:
fix

kuafou commented:
> let me think, without this pr it will not cause actual mamba slot leak, but it will cause check memory failed, do I understand correctly?

Yes.

Without this PR, it’s not necessarily a *real* long-term “slot leak” in the sense of “lost forever,” but it **does create an untracked allocated mamba slot** in the failure path (e.g., `add_one_req()` returns `NO_TOKEN` and scheduling bails out). That slot is:

- **not in** `mamba_pool.free_slots` (because it’s already allocated), and
- **not in** `tree_cache`’s cached mamba values (because the request didn’t finish and the state wasn’t inserted),

so `check_memory()` computes it as `expected - free - cached` and flags it as `leaked_mamba_pages`, which can **trigger the “memory leak detected” crash**.

That’s why the fix explicitly **frees the mamba slot when scheduling fails**. The request stays in `waiting_queue`, and on the next scheduling round it will go through `match_prefix()` again and safely re-allocate (and copy) the mamba state. Keeping the old `mamba_pool_idx` would require teaching `check_memory` to account for “waiting requests holding slots,” which adds complexity for very little gain (since the `copy_from` still happens either way).

kuafou commented:
> > > let me think, without this pr it will not cause actual mamba slot leak, but it will cause check memory failed, do I understand correctly?
> > 
> > 
> > Yes.
> > Without this PR, it’s not necessarily a _real_ long-term “slot leak” in the sense of “lost forever,” but it **does create an untracked allocated mamba slot** in the failure path (e.g., `add_one_req()` returns `NO_TOKEN` and scheduling bails out). That slot is:
> > 
> > * **not in** `mamba_pool.free_slots` (because it’s already allocated), and
> > * **not in** `tree_cache`’s cached mamba values (because the request didn’t finish and the state wasn’t inserted),
> > 
> > so `check_memory()` computes it as `expected - free - cached` and flags it as `leaked_mamba_pages`, which can **trigger the “memory leak detected” crash**.
> > That’s why the fix explicitly **frees the mamba slot when scheduling fails**. The request stays in `waiting_queue`, and on the next scheduling round it will go through `match_prefix()` again and safely re-allocate (and copy) the mamba state. Keeping the old `mamba_pool_idx` would require teaching `check_memory` to account for “waiting requests holding slots,” which adds complexity for very little gain (since the `copy_from` still happens either way).
> 
> I see, I think it will not `lost this slot forwever` since request mamba_pool_idx is not none, and when doing extend slot will be tracked again. But it will cause `check_memory` failed actually.

Yes.

kuafou commented on python/sglang/srt/managers/scheduler.py:
```diff
@@ -2003,6 +2003,25 @@ def _get_new_batch_prefill_raw(
+                # Without this, the slot remains held by a waiting request, causing
+                # check_memory() to detect a "memory leak" and crash the server.
+                # The next schedule round will re-allocate safely via match_prefix().
+                #
```
Yes, sure.

kuafou committed:
refactor

hanming-lu reviewed (commented):
Also I don't see the relationship between https://github.com/sgl-project/sglang/issues/15840 and this PR. https://github.com/sgl-project/sglang/issues/15840 is leaking runtime memory, this PR doesn't touch runtime memory.

kuafou commented on python/sglang/srt/managers/scheduler.py:
```diff
@@ -2100,6 +2102,25 @@ def _get_new_batch_prefill_raw(
+
+        Without this, the slot remains held by a waiting request, causing
+        check_memory() to detect a "memory leak" and crash the server.
+        The next schedule round will re-allocate safely via match_prefix().
```
That is an excellent observation.

The **“Idle”** state in `scheduler.py` is indeed slightly ambiguous, I think. It is entered whenever
`get_next_batch_to_run()` returns `None`, which can happen in two distinct scenarios:

1. **True idle**: the waiting queue is empty.
2. **Resource-blocked**: the waiting queue is non-empty, but available resources (e.g. KV cache or Mamba pool slots) are insufficient to schedule even a single request (for example, when `add_one_req()` returns `AddReqResult.NO_TOKEN`).

In the second case, although there are still pending requests, the scheduler has no runnable work at that moment. As a result, it falls back to `self_check_during_idle()`, which then triggers `check_memory()`.

So the behavior is consistent with the current control flow, but the term **“Idle”** is arguably overloaded here — it also covers a *temporarily resource-blocked* state rather than a strictly empty system.

MegaGriffin commented:
> > Also I don't see the relationship between #15840 and this PR. #15840 is leaking runtime memory, this PR doesn't touch runtime memory.
> 
> Thanks for the clarification.
> 
> You’re right — #15840 refers to a runtime GPU memory leak, and this PR does not directly address that issue. The only relationship is that this bug was found while investigating #15840 locally.
> 
> This PR fixes a separate scheduler failure-path problem: when scheduling fails, a COW-allocated mamba slot is left temporarily untracked, which can cause check_memory() to report a leak and crash the server.
> 
> If it would be clearer, I’m happy to update the PR title and/or Motivation to remove the direct reference to #15840 and instead describe it as “discovered during investigation of #15840.” Please let me know what you prefer.


Hi @kuafou , thanks for your insight. Did your local test crash on check_memory() function with the report of "token_to_kv_pool_allocator memory leak", or it crash on "req_to_token_pool memory leak detected!" ?

kuafou commented:
> > > Also I don't see the relationship between #15840 and this PR. #15840 is leaking runtime memory, this PR doesn't touch runtime memory.
> > 
> > 
> > Thanks for the clarification.
> > You’re right — #15840 refers to a runtime GPU memory leak, and this PR does not directly address that issue. The only relationship is that this bug was found while investigating #15840 locally.
> > This PR fixes a separate scheduler failure-path problem: when scheduling fails, a COW-allocated mamba slot is left temporarily untracked, which can cause check_memory() to report a leak and crash the server.
> > If it would be clearer, I’m happy to update the PR title and/or Motivation to remove the direct reference to #15840 and instead describe it as “discovered during investigation of #15840.” Please let me know what you prefer.
> 
> Hi @kuafou , thanks for your insight. Did your local test crash on check_memory() function with the report of "token_to_kv_pool_allocator memory leak", or it crash on "req_to_token_pool memory leak detected!" ?

Hi, I have the logs from the test and will post them shortly. In the meantime, could you share exactly what the original error message/report was on your side?

kuafou commented:
@MegaGriffin  

Hi, that my test scripts and error log. Hope this helps!


# Server Side Scripts

```
export LD_LIBRARY_PATH=$CONDA_PREFIX/lib:$LD_LIBRARY_PATH
export MODEL_PATH=/mnt/nvme0n1/home/jiaqi/.cache/modelscope/hub/models/Qwen/Qwen3-Next-80B-A3B-Instruct
export MODEL_LENGTH=262144
export TP_SIZE=8
export SERVED_NAME=test_serving

export COMMON_LAUNCH_ARGS="--trust-remote \
    --model-path $MODEL_PATH \
    --context-length ${MODEL_LENGTH} \
    --tp-size ${TP_SIZE} \
    --served-model-name ${SERVED_NAME} \
    --host 0.0.0.0 \
    --grammar-backend outlines \
    --enable-memory-saver \
    --allow-auto-truncate \
    --sampling-backend flashinfer \
    --attention-backend triton \
    --disable-overlap-schedule \
    --mem-fraction-static 0.9 \
    --enable-metrics \
    --log-level info "

    # --attention-backend flashinfer \
python -m sglang.launch_server ${COMMON_LAUNCH_ARGS}
```

#  Batchmark Scripts

```
python3 -m sglang.bench_serving \
    --model /mnt/nvme0n1/home/jiaqi/.cache/modelscope/hub/models/Qwen/Qwen3-Next-80B-A3B-Instruct \
    --backend sglang \
    --host 127.0.0.1 \
    --port 30000 \
    --warmup-requests 0 \
    --dataset-name generated-shared-prefix \
    --gsp-num-groups 1 \
    --gsp-prompts-per-group 50 \
    --gsp-system-prompt-len 150000 \
    --gsp-question-len 400 \
    --gsp-output-len 8192
```



# Error Log

```
[2026-01-12 10:57:27 TP0] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 7819, #new_token_ratio: 0.6485 -> 0.9570
[2026-01-12 10:57:27 TP1] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 7819, #new_token_ratio: 0.6485 -> 0.9570
[2026-01-12 10:57:27 TP6] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 7819, #new_token_ratio: 0.6485 -> 0.9570
[2026-01-12 10:57:27 TP3] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 7819, #new_token_ratio: 0.6485 -> 0.9570
[2026-01-12 10:57:27 TP4] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 7819, #new_token_ratio: 0.6485 -> 0.9570
[2026-01-12 10:57:27 TP2] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 7819, #new_token_ratio: 0.6485 -> 0.9570
[2026-01-12 10:57:27 TP5] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 7819, #new_token_ratio: 0.6485 -> 0.9570
[2026-01-12 10:57:27 TP7] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 7819, #new_token_ratio: 0.6485 -> 0.9570
[2026-01-12 10:57:27 TP0] Decode batch, #running-req: 28, #full token: 388852, full token usage: 0.98, mamba num: 57, mamba usage: 0.13, cuda graph: True, gen throughput (token/s): 673.12, #queue-req: 22, 
[2026-01-12 10:57:29 TP0] Decode batch, #running-req: 28, #full token: 389972, full token usage: 0.98, mamba num: 57, mamba usage: 0.13, cuda graph: True, gen throughput (token/s): 669.53, #queue-req: 22, 
[2026-01-12 10:57:31 TP0] Decode batch, #running-req: 28, #full token: 391092, full token usage: 0.99, mamba num: 57, mamba usage: 0.13, cuda graph: True, gen throughput (token/s): 673.28, #queue-req: 22, 
[2026-01-12 10:57:32 TP0] Decode batch, #running-req: 28, #full token: 392212, full token usage: 0.99, mamba num: 57, mamba usage: 0.13, cuda graph: True, gen throughput (token/s): 673.00, #queue-req: 22, 
[2026-01-12 10:57:34 TP0] Decode batch, #running-req: 28, #full token: 393332, full token usage: 0.99, mamba num: 57, mamba usage: 0.13, cuda graph: True, gen throughput (token/s): 671.93, #queue-req: 22, 
[2026-01-12 10:57:36 TP0] Decode batch, #running-req: 28, #full token: 394452, full token usage: 0.99, mamba num: 57, mamba usage: 0.13, cuda graph: True, gen throughput (token/s): 670.30, #queue-req: 22, 
[2026-01-12 10:57:37 TP0] Decode batch, #running-req: 28, #full token: 395572, full token usage: 1.00, mamba num: 57, mamba usage: 0.13, cuda graph: True, gen throughput (token/s): 668.98, #queue-req: 22, 
[2026-01-12 10:57:39 TP0] Decode batch, #running-req: 28, #full token: 396692, full token usage: 1.00, mamba num: 57, mamba usage: 0.13, cuda graph: True, gen throughput (token/s): 668.86, #queue-req: 22, 
[2026-01-12 10:57:39 TP0] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 8113, #new_token_ratio: 0.6631 -> 0.9929
[2026-01-12 10:57:39 TP2] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 8113, #new_token_ratio: 0.6631 -> 0.9929
[2026-01-12 10:57:39 TP4] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 8113, #new_token_ratio: 0.6631 -> 0.9929
[2026-01-12 10:57:39 TP5] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 8113, #new_token_ratio: 0.6631 -> 0.9929
[2026-01-12 10:57:39 TP7] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 8113, #new_token_ratio: 0.6631 -> 0.9929
[2026-01-12 10:57:39 TP1] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 8113, #new_token_ratio: 0.6631 -> 0.9929
[2026-01-12 10:57:39 TP6] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 8113, #new_token_ratio: 0.6631 -> 0.9929
[2026-01-12 10:57:39 TP3] KV cache pool is full. Retract requests. #retracted_reqs: 1, #new_tokens_gained: 8113, #new_token_ratio: 0.6631 -> 0.9929
[2026-01-12 10:57:41 TP0] Decode batch, #running-req: 27, #full token: 389237, full token usage: 0.98, mamba num: 55, mamba usage: 0.12, cuda graph: True, gen throughput (token/s): 631.80, #queue-req: 23, 
[2026-01-12 10:57:43 TP2] Scheduler hit an exception: Traceback (most recent call last):
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 2996, in run_scheduler_process
    scheduler.event_loop_normal()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 1105, in event_loop_normal
    self.self_check_during_idle()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 330, in self_check_during_idle
    self.check_memory()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 244, in check_memory
    raise_error_or_warn(
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/utils/common.py", line 3774, in raise_error_or_warn
    raise ValueError(message)
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}


[2026-01-12 10:57:43] SIGQUIT received. signum=None, frame=None. It usually means one child failed.
[2026-01-12 10:57:43 TP1] Scheduler hit an exception: Traceback (most recent call last):
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 2996, in run_scheduler_process
    scheduler.event_loop_normal()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 1105, in event_loop_normal
    self.self_check_during_idle()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 330, in self_check_during_idle
    self.check_memory()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 244, in check_memory
    raise_error_or_warn(
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/utils/common.py", line 3774, in raise_error_or_warn
    raise ValueError(message)
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}


[2026-01-12 10:57:43 TP5] Scheduler hit an exception: Traceback (most recent call last):
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 2996, in run_scheduler_process
    scheduler.event_loop_normal()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 1105, in event_loop_normal
    self.self_check_during_idle()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 330, in self_check_during_idle
    self.check_memory()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 244, in check_memory
    raise_error_or_warn(
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/utils/common.py", line 3774, in raise_error_or_warn
    raise ValueError(message)
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}


[2026-01-12 10:57:43 TP6] Scheduler hit an exception: Traceback (most recent call last):
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 2996, in run_scheduler_process
    scheduler.event_loop_normal()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 1105, in event_loop_normal
    self.self_check_during_idle()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 330, in self_check_during_idle
    self.check_memory()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 244, in check_memory
    raise_error_or_warn(
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/utils/common.py", line 3774, in raise_error_or_warn
    raise ValueError(message)
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}


[2026-01-12 10:57:43 TP3] Scheduler hit an exception: Traceback (most recent call last):
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 2996, in run_scheduler_process
    scheduler.event_loop_normal()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 1105, in event_loop_normal
    self.self_check_during_idle()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 330, in self_check_during_idle
    self.check_memory()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 244, in check_memory
    raise_error_or_warn(
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/utils/common.py", line 3774, in raise_error_or_warn
    raise ValueError(message)
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}


[2026-01-12 10:57:43 TP4] Scheduler hit an exception: Traceback (most recent call last):
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 2996, in run_scheduler_process
    scheduler.event_loop_normal()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 1105, in event_loop_normal
    self.self_check_during_idle()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 330, in self_check_during_idle
    self.check_memory()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 244, in check_memory
    raise_error_or_warn(
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/utils/common.py", line 3774, in raise_error_or_warn
    raise ValueError(message)
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}


[2026-01-12 10:57:43 TP0] Scheduler hit an exception: Traceback (most recent call last):
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 2996, in run_scheduler_process
    scheduler.event_loop_normal()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 1105, in event_loop_normal
    self.self_check_during_idle()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 330, in self_check_during_idle
    self.check_memory()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 244, in check_memory
    raise_error_or_warn(
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/utils/common.py", line 3774, in raise_error_or_warn
    raise ValueError(message)
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}


[2026-01-12 10:57:43 TP7] Scheduler hit an exception: Traceback (most recent call last):
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 2996, in run_scheduler_process
    scheduler.event_loop_normal()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler.py", line 1105, in event_loop_normal
    self.self_check_during_idle()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 330, in self_check_during_idle
    self.check_memory()
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/managers/scheduler_runtime_checker_mixin.py", line 244, in check_memory
    raise_error_or_warn(
  File "/mnt/nvme0n1/home/jiaqi/workspaces/opensources/sglang/python/sglang/srt/utils/common.py", line 3774, in raise_error_or_warn
    raise ValueError(message)
ValueError: token_to_kv_pool_allocator memory leak detected! full_available_size=6016, full_evictable_size=390685, self.token_to_kv_pool_allocator.size=396701, self.tree_cache.full_protected_size()=0
mamba_available_size=315, mamba_evictable_size=138, self.req_to_token_pool.mamba_pool.size=454, self.tree_cache.mamba_protected_size()=0, leaked_full_pages=None, leaked_mamba_pages={191}


./outputs/run-mml.sh: line 24: 3557732 Killed                  python -m sglang.launch_server ${COMMON_LAUNCH_ARGS}
```

hanming-lu reviewed (approved):
This is a mitigation where server incorrectly thinks the server is idle, triggering memory check. From the server logs, it looks to happen during retractions.

Please update the description, thanks.
https://github.com/sgl-project/sglang/pull/16067#discussion_r2669989038

yizhang2077 committed:
remove conflict tests

yizhang2077 commented:
/rerun-failed-ci
